### PR TITLE
Feature: query filter

### PR DESCRIPTION
**Pull request by Zorin95670**
Add mechanisme to filter data from query.

### CODE REVIEW

**Review by aboubakar14 — COMMENTED** [review]
Your 1st commit is not (only) what it says. Split it
Moreover, you deleted some "Vincent Moittie" in the next commit instead of not having them at all in the commits

File: src/main/java/com/ditrit/letomodelizerapi/config/Constants.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
+package com.octo.model.common;
+
+/**
+ * All constants value.
+ *
+ * @author Vincent Moittié
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/config/Constants.java]: done

[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/config/Constants.java]: done

File: src/main/java/com/ditrit/letomodelizerapi/controller/model/QueryFilter.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    /**
+     * The maximum size of resources.
+     */
+    @QueryParam("count")
+    @Min(1)
+    @Max(Constants.MAXIMUM_RESOURCE_SIZE)
+    private Integer count = Constants.DEFAULT_RESOURCE_SIZE;
+
+    /**
+     * Name of field to sort.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/controller/model/QueryFilter.java:41]: done

File: src/main/java/com/ditrit/letomodelizerapi/controller/model/QueryFilter.java [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+import javax.validation.constraints.Max;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.Pattern;
+import javax.ws.rs.QueryParam;
+import java.util.Optional;
+
+/**
+ * Class to manage query option for controller.
+ *
+ * @author Vincent Moittié
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/controller/model/QueryFilter.java]: done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/PredicateOperator.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
+package com.ditrit.letomodelizerapi.persistence.specification;
+
+import java.util.Arrays;
+import java.util.Optional;
+
+/**
+ * List of available operator for QueryFilter.
+ *
+ * @author Vincent Moittié
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/PredicateOperator.java]: done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/DatePredicateFilter.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        }
+    }
+
+    @Override
+    public final boolean extract() {
+        if (this.getValues().length == 0) {
+            return false;
+        }
+
+        for (int index = 0; index < this.getValues().length; index++) {
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/DatePredicateFilter.java:116]: done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/FilterType.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
+package com.ditrit.letomodelizerapi.persistence.specification.filter;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation to indicate if field is filtering.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/FilterType.java:9]: Done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/NumberPredicateFilter.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+            if (this.getIsNotOperator(index)) {
+                allNotEquals += 1;
+            } else {
+                allEquals += 1;
+            }
+        }
+
+        if (this.getValues().length == allNotEquals) {
+            return builder
+                    .not(root.get(this.getName()).in(Arrays.stream(this.getValues()).map(Long::parseLong).toList()));
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/NumberPredicateFilter.java:60]: done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+
+import javax.persistence.criteria.CommonAbstractCriteria;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.Expression;
+import javax.persistence.criteria.From;
+import javax.persistence.criteria.Predicate;
+import java.util.Arrays;
+
+/**
+ * Abstract class to implements default field for QueryFilter.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java:15]: done

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+     */
+    protected static final String NOT_DELIMITER = "not_";
+    /**
+     * Operator of this filter.
+     */
+    private PredicateOperator[] operators;
+    /**
+     * Indicate if it's not operator.
+     */
+    private boolean[] isNotOperators;
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java:35]: no

File: src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+    public void setValues(final String value) {
+        if (value != null) {
+            this.values = value.split(OR_DELIMITER);
+        } else {
+            this.values = new String[0];
+        }
+    }
+
+    /**
+     * Get operator.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/persistence/specification/filter/PredicateFilter.java:133]: sure.

File: src/main/java/com/ditrit/letomodelizerapi/reflect/FieldUtils.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
+package com.ditrit.letomodelizerapi.reflect;
+
+import java.lang.reflect.Field;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Provide some helpful method to work with field.
```
[the remark this answers is not in the document]
Comment (Zorin95670) [reply src/main/java/com/ditrit/letomodelizerapi/reflect/FieldUtils.java:10]: done

**Review by aboubakar14 — APPROVED** [review]

**Pull request merged**